### ISSUE TITLE
Several buffer overflows in `str2net`

### ISSUE THREAD

**Issue by hg** [gsliepen/tinc#292]
I was doing a bit of an experiment with fuzzing some internal functions that receive input from config files.

Here is a simple program to test `str2net` (replace `main` in `tincctl.c` with it):

```c
#include "system.h"
#include "subnet.h"

int main() {
    char data[1024];
    size_t nread = fread(data, 1, sizeof(data) - 1, stdin);

    if (nread <= 0) {
        return 1;
    }
    if (data[nread - 1] == '\n') {
        --nread;
    }
    data[nread] = 0;

    subnet_t sub = {0};
    if (!str2net(&sub, data)) {
        return 1;
    }

    return 0;
}
```

and run it thusly (after building the project with `-fsanitize=address`):

    $ ./src/tinc < test_case_file

or you can [install AFL](https://github.com/google/AFL) and try the full experience (throw a couple of valid and invalid subnets into separate files in `fuzz/in`):

```shell
export CC=afl-gcc
autoreconf -fsi && ./configure
AFL_USE_ASAN=1 make -j
afl-fuzz -m 500000000000 -i fuzz/in/ -o fuzz/out/ ./src/tinc
```

It should quickly find the first overflow (try the test case `00` attached to this message):

```
==2971195==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7ffe80e25410 at pc 0x7f31328b19fb bp 0x7ffe80e24f00 sp 0x7ffe80e246a8
WRITE of size 1010 at 0x7ffe80e25410 thread T0
    #0 0x7f.. in __interceptor_memmove /build/gcc/src/gcc/libsanitizer/sanitizer_common/sanitizer_common_interceptors.inc:810
    #1 0x55.. in str2net src/subnet_parse.c:340
    #2 0x55.. in main src/fuzz_fsck.c:17
    #3 0x7f.. in __libc_start_main (/usr/lib/libc.so.6+0x27b24)
    #4 0x55.. in _start (src/tinc+0x194fd)
```

This can easily be triggered by adding this "subnet" to `tinc.conf` and starting tincd as usual:

```
Subnet = ::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
```

After fixing it with a quick-and-dirty length check

```c
const size_t shift = zero_group_count * 2 - 1;
const size_t double_colon_len = strlen(double_colon);
if (double_colon_len < shift) {
    return false;
}
memmove(double_colon + shift, double_colon, double_colon_len - shift + 1);
```

the fuzzer finds another overflow (try file `01`)

```
==2970727==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7fffdb7c1510 at pc 0x7f42e103b72b bp 0x7fffdb7c1000 sp 0x7fffdb7c07a8
READ of size 1024 at 0x7fffdb7c1510 thread T0
    #0 0x7f.. in __interceptor_strlen /build/gcc/src/gcc/libsanitizer/sanitizer_common/sanitizer_common_interceptors.inc:389
    #1 0x56.. in str2net src/subnet_parse.c:355
    #2 0x56.. in main src/fuzz_fsck.c:17
    #3 0x7f.. in __libc_start_main (/usr/lib/libc.so.6+0x27b24)
    #4 0x56.. in _start (src/tinc+0x194fd)
```

and then another one (try file `02`)

```
==2970892==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7ffd92fa8fd0 at pc 0x7fd0afa2972b bp 0x7ffd92fa8ac0 sp 0x7ffd92fa8268
READ of size 1025 at 0x7ffd92fa8fd0 thread T0
    #0 0x7f.. in __interceptor_strlen /build/gcc/src/gcc/libsanitizer/sanitizer_common/sanitizer_common_interceptors.inc:389
    #1 0x55.. in str2net src/subnet_parse.c:358
    #2 0x55.. in main src/fuzz_fsck.c:17
    #3 0x7f.. in __libc_start_main (/usr/lib/libc.so.6+0x27b24)
    #4 0x55.. in _start (src/tinc+0x194fd)
```

Instead of adding buffer size checks to all of these, and seeing that we have a large buffer to work with, and the longest *valid* address we would ever see (full-length IPv6 with `/128` and maximum possible weight (`2 ** 63 - 1` on some weird architecture?)) is on the order of 100 bytes, the simplest way to fix this is to add a simple check like so:

```c
bool str2net(subnet_t *subnet, const char *subnetstr) {
	char str[1024];
	strncpy(str, subnetstr, sizeof(str));
	str[sizeof(str) - 1] = 0;

	// The longest valid address we could ever see would be on the order of 100 chars.
	// This check prevents several buffer overflows below with specially crafted subnets.
	if (strlen(str) > sizeof(str) / 2) {
	    return false;
	}

	// ...
}
```

After adding this check and running AFL for several hours, I haven't been able to find anything else in this function. With the config line above, tincd now prints the expected message about an invalid subnet instead of crashing.

Any thoughts?

---

[Crashing inputs](https://github.com/gsliepen/tinc/files/6878141/testcases.zip)

**Comment by gsliepen** [comment]
I made it check the position of the last colon instead in commit 04b588ea6be2ebd62922b638eae90a6a47326370. I think that fixes it as well; all the string manipulation depends on that. I noticed that `tinc add Subnet` warned about the failing test case being an invalid Subnet, but it added it anyway due to a missing `return` statement. I fixed that, and added some tests for malformed Subnets to `variables.test` in commit 743d3cefc244e669c9265da8355e8722d990fc8e. This still needs to be backported to 1.0.

Instead of writing your own `main()` to call `str2net`, we could instead just let AFL fuzz `tincd` itself. You can set `DeviceType = dummy`, and have a `tinc-up` script shut down tinc immediately after it started.

**Comment by gsliepen** [comment]
I see also need to start creating pull requests for myself so I don't push fixes that don't pass CI. I missed a `git add` and even though I ran the test suite through ASAN myself with success, there apparently are subtle differences between GCC's and Clang's sanitizers.

**Comment by hg** [comment]
Thanks a lot for such quick response. It seems to be working fine in my testing.

I'll look into full fuzzing a bit later, this was actually the first step in that direction.

>I see also need to start creating pull requests for myself so I don't push fixes that don't pass CI

You can simply push to a different branch (any will do) and see if it breaks. If CI checks pass fine, push the same commit to `1.1`.

**Comment by hg** [comment]
`afl++` found more overflows.

[testcases.zip](https://github.com/gsliepen/tinc/files/6884135/testcases.zip)

**Comment by gsliepen** [comment]
Hmkay I think I'm going to completely rewrite `subnet_parse()` to avoid any string manipulation. The code has been like this because in the past, not all operating systems supported IPv6, but now I think it should be possible to just use `inet_ntop()` for IPv4 and IPv6 Subnets, and a simple `sscanf()` for the MAC Subnets.

**Comment by gsliepen** [comment]
https://github.com/gsliepen/tinc/tree/fix/subnet-parsing

**Comment by hg** [comment]
The latest solution looks great. There probably wasn't much sense in doing it anymore (system libraries are very well tested anyway), but I ran about 50k cycles of `afl++` on the latest version, which is much more than was required to trigger previous overflows, and it didn't find any problems.

**Comment by gsliepen** [comment]
Thanks. Fixed in commits f7f7d8b15e12471c95bbd291195e83c5126a0631 (1.0) and eda0cc6243c5292fb35fdadf6ec130983c530b2c (1.1).
